doc/radosgw/s3select.rst

```diff
@@ -0,0 +1,265 @@
+===============
+ Ceph s3 select 
+===============
+
+.. contents::
+
+Overview
+--------
+
+    | The purpose of **s3 select** engine is to create an efficient pipe between user client to storage node (the engine should be close as possible to storage).
+    | It enables the user to define the exact portion of data should be received by his side.
+    | It also enables for higher level analytic-applications (such as SPARK-SQL) , using that feature to improve their latency and throughput.
+
+    | For example, a s3-object of several GB (CSV file), a user needs to extract a single column which filtered by another column.
+    | As the following query:
+    | ``select customer-id from s3Object where age>30 and age<65;``
+
+    | Currently the whole s3-object must retrieve from OSD via RGW before filtering and extracting data.
+    | By "pushing down" the query into OSD , it's possible to save a lot of network and CPU(serialization / deserialization).
+
+    | **The bigger the object, and the more accurate the query, the better the performance**.
+ 
+Basic workflow
+--------------
+    
+    | S3-select query is sent to RGW via `AWS-CLI <https://docs.aws.amazon.com/cli/latest/reference/s3api/select-object-content.html>`_
```

Comment:
we should give an aws-cli example for easy testing, but we should not state that this is required for the basic workflow.
please see how we document other S3 commands.
usually we put boto3 and awc-cli examples [here](https://github.com/ceph/ceph/tree/master/examples/boto3), but, it does make sense to embed that in the documentation as well.

Reply:
(I strongly support having examples in our doc, but agree on the language.  awscli is not the only or best way to use s3 select--though it looks like it may be the easiest to play with)

Reply:
current implementation support **ONLY** AWS-CLI , it explicitly extract certain tags from the incoming message, and also formatting the response message for AWS-cli (header-type,message-type,crc verification).
AWS-cli is well defined (including v4 signature), and widely used (boto3).

before defining(and implementing) other types of http-clients, we should define the purpose of it.
ceph-s3-select should support presto,hive,spark , probably similar to the way aws-s3-select is doing that.
supporting those analytical tools is highly important.

Reply:
> before defining(and implementing) other types of http-clients, we should define the purpose of it.

there is no need to define the types of http clients. all clients that we support (including boto3 and aws-cli) are http clients, and they all support the same http format both in the requests and the responses.
the purpose is simple, if we support the standard format (used by aws-cli) we support any client that uses this format, regardless of programming language of framework it uses.
for testing purposes we may say that boto3 (and aws-cli , which are actually the same code) will be used, but we **should not** make any assumptions in the code that these are the only clients being used.

Reply:
ok, as discussed 
current implementation support any http-client which comply to aws cli specific interface, and should be able to handle the current formmated response.